### PR TITLE
VB-4332 - Email notifications for public users with gov notify

### PR DESCRIPTION
**Pull request by adaviesMOJ**
## What does this PR do?
Introduce sending email to for booked visits to main contact if they provide an email address.

## Dev work
- Add new clients for prisoner-contact-registry and prisoner search to help get all information needed to make email template.
- Add email service, email templates and email configs to enable configuring and sending emails via gov notify.
- Add integration test to assert functionality is working.

Note: In prod, email sending var has been set to false for now.

## JIRA Link
https://dsdmoj.atlassian.net/browse/VB-4332

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -41,4 +41,4 @@ The `rename-project.bash` script takes a single argument - the name of the proje
 * The project description (class name with spaces between the words)
 * The main package name (project name with hyphens removed)
 
-It then performs a search and replace and directory renames so the project is ready to be used.
\ No newline at end of file
+It then performs a search and replace and directory renames so the project is ready to be used.
```
Comment (dhirajshettymoj) [inline README.md:44]: Not part of this ticket. Just an FYI - have created a Tech Debt ticket https://dsdmoj.atlassian.net/browse/VB-4856 to update README

File: src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/EmailSenderService.kt [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    }
+  }
+
+  private fun handleBookedEvent(visit: VisitDto): SendEmailNotificationDto {
+    val templateVars = mutableMapOf(
+      "ref number" to visit.reference,
+      "prison" to prisonRegisterService.getPrisonName(visit.prisonCode),
+      "time" to dateUtils.getFormattedTime(visit.startTimestamp.toLocalTime()),
+      "end time" to dateUtils.getFormattedTime(visit.endTimestamp.toLocalTime()),
+      "arrival time" to "45",
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/EmailSenderService.kt:64]: should we get this from a service with a hardcoded value of 45 later update the service to get it from the visit-scheduler?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/EmailSenderService.kt [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+    val templateName = EmailTemplateNames.VISIT_BOOKING
+
+    return SendEmailNotificationDto(templateName = templateName, templateVars = templateVars)
+  }
+
+  private fun getPrisoner(visit: VisitDto): String {
+    return prisonerSearchService.getPrisoner(visit.prisonerId)?.let {
+      it.firstName + " " + it.lastName
+    } ?: "Prisoner"
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/EmailSenderService.kt:82]: suggestion only, can be ignored - could do a toString() like you did below

File: src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/SmsSenderService.kt [hunk 4]
```diff
@@ -83,14 +83,13 @@ class SmsSenderService(
...
     val templateVars = mutableMapOf(
-      "prison" to prisonName,
+      "prison" to prisonRegisterService.getPrisonName(visit.prisonCode),
       "time" to dateUtils.getFormattedTime(visit.startTimestamp.toLocalTime()),
       "dayofweek" to dateUtils.getFormattedDayOfWeek(visit.startTimestamp.toLocalDate()),
       "date" to dateUtils.getFormattedDate(visit.startTimestamp.toLocalDate()),
     )
+
     return templateVars
   }
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/SmsSenderService.kt:94]: just a thought should this move to a different service that can be used by both EmailSenderService and SmsSenderService?

**Review by dhirajshettymoj — COMMENTED** [review]
Nice work - just some minor comments added. Can be ignored

File: src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/SmsSenderService.kt (hunk 4 above)
Comment (adaviesMOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/SmsSenderService.kt:94]: After discussing, we will leave this pull request as is. Then after merging, we'll work on splitting each email template into their own service.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/EmailSenderService.kt (hunk 3 above)
Comment (adaviesMOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/EmailSenderService.kt:82]: This already returns a string, im not 100% sure what you mean, could you put a code snippet here as example?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/EmailSenderService.kt (hunk 2 above)
Comment (adaviesMOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/EmailSenderService.kt:64]: For now, it's unlikely to be changed for a long time. I'm happy to leave it for now 😎

File: src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/EmailSenderService.kt (hunk 3 above)
Comment (adaviesMOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/EmailSenderService.kt:82]: After discussion i understand overriding the toString() call on the Dto returned is what was suggested. I'll implement this and chase up what the "backup" value should be if the search results in null

**Review by dhirajshettymoj — APPROVED** [review]
LGTM

File: src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/dto/prisoner/search/PrisonerSearchResultDto.kt [hunk 5]
```diff
@@ -11,6 +11,6 @@ data class PrisonerSearchResultDto(
   val lastName: String,
 ) {
   override fun toString(): String {
-    return "$firstName $lastName"
+    return "Your visit to see $firstName $lastName"
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/dto/prisoner/search/PrisonerSearchResultDto.kt:14]: sorry I know I suggested this but given that we are now appending "Your visit to see" could we move it into a different method in the service?

**Review by dhirajshettymoj — COMMENTED** [review]
1 small comment

**Review by dhirajshettymoj — APPROVED** [review]
LGTM

**Pull request merged**